src/prognostic_equations/implicit/autodiff_utils.jl

```diff
@@ -0,0 +1,92 @@
+function scalar_level_index_pairs(field_vector)
+    scalar_names = scalar_field_names(field_vector)
+    scalar_index_and_level_pairs =
+        unrolled_map(enumerate(scalar_names)) do (scalar_index, name)
+            field = MatrixFields.get_field(field_vector, name)
+            is_one_level = field isa Fields.SpectralElementField
+            Iterators.map(
+                level_index -> (scalar_index, level_index),
+                Base.OneTo(is_one_level ? 1 : Spaces.nlevels(axes(field))),
+            )
+        end
+    return enumerate(Iterators.flatten(scalar_index_and_level_pairs))
+end
+
+# A view of the data for one point in a field, which can be used like a Ref.
+Base.@propagate_inbounds function point(field, level_index, column_index...)
+    column_field = Fields.column(field, column_index...)
+    column_field isa Fields.PointField && return column_field
+    level_index_offset = Operators.left_idx(axes(column_field)) - 1
+    return Fields.level(column_field, level_index + level_index_offset)
+end
+
+# TODO: This needs to be moved into ClimaCore to avoid breaking precompilation.
+# function ClimaComms.context(x::Fields.FieldVector)
+#     values = Fields._values(x)
+#     isempty(values) && error("Empty FieldVector has no device or context")
+#     index = unrolled_findfirst(Base.Fix2(!isa, Fields.PointField), values)
+#     return ClimaComms.context(values[isnothing(index) ? 1 : index])
+# end
+
+##################################################
+## Nonstandard rules for automatic differentiation
+##################################################
+
+# Every nonstandard rule should be defined using a method that specializes on
+# the tag Jacobian. Not specializing on any tag overwrites the generic method
+# for Dual in ForwardDiff and breaks precompilation, while specializing on the
+# default tag Nothing causes the type piracy test to fail.
+
+# Set the derivative of sqrt(x) to iszero(x) ? zero(x) : inv(2 * sqrt(x)) in
+# order to properly handle derivatives of x * sqrt(x). Without this change, the
+# derivative of the turbulent kinetic energy dissipation tendency, which is a
+# linear function of ᶜtke⁰ * sqrt(ᶜtke⁰), is NaN whenever tke⁰ = 0 at any point.
+# In general, this change is valid when every function of sqrt(x) can be
+# expanded around x = 0 with a leading term of the form c * x^p * sqrt(x), where
+# c and p are constants and p ≥ -1/2. For example, this will be the case for any
+# linear function of sqrt(x), and more generally any linear function of
+# f(x) * sqrt(x), where f(x) has a Taylor series expansion around x = 0. On the
+# other hand, this change will not be valid for a function like sqrt(x) / x,
+# since its derivative at x = 0 should be ±Inf but this change makes it NaN. If
+# a function whose leading term has p < -1/2 is ever called by
+# set_implicit_precomputed_quantities! or implicit_tendency!, our strategy for
+# avoiding NaNs in the dissipation tendency derivative will need to be modified.
```

Comment:
I think it's valid for p >= +1/2. The derivative of x^p * sort(x) is C * x^p / sort(x) which is ±Inf for p < 1/2.

Reply:
Good catch! I will update it.